AG-10326 Set default bullet colorRange based on user-defined series fill

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-10326

olegat committed:
CRT-190 Set default bullet colorRange based on user-defined series fill

olegat committed:
AG-10326 Fix lint

olegat committed:
AG-10326 nx format

olegat committed:
AG-10326 Fix sonar error

olegat commented on packages/ag-charts-enterprise/src/series/bullet/bulletSeriesProperties.ts:
```diff
@@ -103,4 +102,7 @@ export class BulletSeriesProperties extends AbstractBarSeriesProperties<AgBullet
     readonly tooltip = new SeriesTooltip<AgBulletSeriesTooltipRendererParams>();
+
+    @Validate(COLOR_STRING) // Internal: Set by paletteFactory.
+    backgroundFill: CssColor = 'white';
```
This isn't a good solution IMHO. Is there an easy to retrieve the `themeTemplatePalette` from the module's palette factory?

alantreadway replied:
The WaterfallSeries has some special handling in this respect, does this look relevant?
https://github.com/ag-grid/ag-charts/blob/65c5647df895858398b131ee07ac30e28c7eb12b/packages/ag-charts-enterprise/src/series/waterfall/waterfallModule.ts#L19-L46

olegat replied:
Doesn't look relevant. The `positive`, `negative` and `total` options are documented in the API: https://charts.ag-grid.com/options/series/waterfall/#reference-AgWaterfallSeriesOptions-item

The issue that I'm facing here is that `backgroundFill` is not documented and is not meant to be. It's only used for internal implementation in _bulletSeries.ts_. The user is not meant to be able to reconfigured `backgroundFill`, but they still can simply by set `options.series[].backgroundFill`

alantreadway replied:
Discussed offline, I've got the wrong end of the stick

olegat committed:
AG-10326 Add OptionalPropertiesArray validation type

This fixes a test regression where no warnings are shown when
`series[].colorRanges[].stop` is set to a negative number.

alantreadway commented on packages/ag-charts-community/src/util/properties.ts:
```diff
@@ -72,6 +73,58 @@ export class PropertiesArray<T extends BaseProperties> extends Array {
     }
 }
 
+export class OptionalPropertiesArray<T extends BaseProperties> {
```
Maybe sync with @iMoses about the use-case for this?

olegat replied:
Resolved. We've decided to use an empty `PropertiesArray` instance instead.

olegat committed:
Revert "AG-10326 Add OptionalPropertiesArray validation type"

This reverts commit 62bb21757913f5bfd004fedff2b3c7dcbbc20882.

olegat committed:
AG-10326 Use an empty colorRange as the default

The colorRange property need to be set to a PropertiesArray so it can
validate the color and stop properties. However, the "default color range"
uses an interpolation of the background Fill and series fill, so it
cannot be calculated at compile-time.

Therefore, we'll use an empty colorRange to mean "interpolate". As a
consequence though, no warning message is shown when colorRange is
explicitly set to [].

alantreadway reviewed (approved):
LGTM
